language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+    withBinds = flip . foldr $ withBind
+
+    checkWellformBinds :: NonEmpty (Binding TyName Name uni ann) -> TypeCheckM uni ann ()
+    checkWellformBinds = traverse_ $ checkWellformBind Rec
+
+    assertNoRecConflict :: [T.Text] -> TypeCheckM uni ann ()
+    assertNoRecConflict = maybe (pure ())
+                          (throwError . ConflictingRecBinds ann1)
+                           . hasDuplicate
+
+    -- TODO: duplicate code, also move to PlutusIR.hs
+    bindingNewIds :: Binding TyName Name uni ann -> [T.Text]
+    bindingNewIds = \case
+        TermBind _ _ d _ -> [nameString $ varDeclName d]
+        DatatypeBind _ (Datatype _ _ _ des vdecls) ->
+            nameString des : fmap (nameString . varDeclName) vdecls
+        _typebind -> []
+
+    -- a binding may introduce max. 1 new id in scope, hence maybe
+    -- TODO: duplicate code, also move to PlutusIR.hs
+    bindingNewTyId :: Binding TyName Name uni ann -> Maybe T.Text
+    bindingNewTyId = \case
+        TypeBind _ tvdecl _ ->
+            pure . nameString . unTyName $ tyVarDeclName tvdecl
+        DatatypeBind _ (Datatype _ tvdecl _ _ _) ->
+            pure . nameString . unTyName $ tyVarDeclName tvdecl
+        _termbind -> mempty
+
+-- See the [Global uniqueness] and [Type rules] notes.
+-- | Check a 'Term' against a 'NormalizedType'.
+checkTypeM
+    :: (GShow uni, GEq uni, DefaultUni <: uni)
+    => ann -> Term TyName Name uni ann -> Normalized (Type TyName uni ()) -> TypeCheckM uni ann ()
+
+-- [infer| G !- term : vTermTy]    vTermTy ~ vTy
+-- ---------------------------------------------
+-- [check| G !- term : vTy]
+checkTypeM ann term vTy = do
+    vTermTy <- inferTypeM term
+    when (vTermTy /= vTy) $ throwError (TypeMismatch ann (void term) (unNormalized vTermTy) vTy)
+
+-- | Extends the environment(s) with a bind.
+-- NOTE: there are no well-formed or duplication checks here, it just adds to the environment(s) like withVar/withTyVar do.
+-- For checks see `checkWellformBind`.
+withBind :: forall uni ann a.
+           Binding TyName Name uni ann
+         -> TypeCheckM uni ann a
+         -> TypeCheckM uni ann a
+withBind b m = case b of
+  TermBind _ _ (VarDecl _ n ty) _rhs -> do
+      -- OPTIMIZE: redundant, usually this function is called together with `checkWellformBind`,
+      -- which performs normalization there as well
+      vTy <- normalizeTypeM $ void ty
+      withVar n vTy m
+  TypeBind _ (TyVarDecl _ tn k) _rhs ->
+      withTyVar tn (void k) m
+  DatatypeBind _ dt@(Datatype ann (TyVarDecl _ tn k) tyargs des vdecls) -> do
+      desType <- mkDestructorType
+      -- normalize all data-constructors's types.
+      normConstrs <- for vdecls $ \(VarDecl _ n ty) ->
+                                     -- make sure to add the implicitly-qualified tyargs foralls for each constructor.
+                                     (n,) <$> (normalizeTypeM $ void $ PIR.mkIterTyForall tyargs ty)
```

Comment:
@michaelpj @effectfully I pushed some new code. I used the `rename` manually applied here and there, instead of using `normalizeType`, because I didn't want to "unnecessarily" call normalize too much. Can you take a look if this solves the problem of global uniqueness preservation?